File: fix_v323_final.js
const fs = require('fs');
const p = 'C:/Users/Server/.openclaw/workspace/Shop_vnext/src/Index.html';
let c = fs.readFileSync(p, 'utf8');

// The core issue:
// h+="<div class='b2 bx' onclick=...">..."
// If I change outer " to ', I must ALSO change class='...' to class="..."
// And fix the onclick quoting.
//
// Simplest approach: use regex to fix ALL broken onclick patterns
// regardless of location.
//
// Broken patterns (all inside double-quoted JS strings starting with h+="):
// 1. onclick=\"func('"+VAL+"')\">
// 2. onclick="func('"+VAL+"')\">  (if the backslash was removed)
// 3. onclick="func('"+VAL+"','"+VAL2+"')\">

// Pattern matches: onclick followed by = then either \" or " then func name
// Then single-quote then "+VAL+" then single-quote then ) then quote then >

// Use a broad regex approach: find all instances and fix them one by one

let fixes = 0;

// Regex: find onclick="..." or onclick=\"..." patterns where the attribute value
// starts with funcName('"+
// Captures: funcName and the expression inside

// For patterns like: onclick="cSt('"+id+"','รอดำเนินการ')"
// Or: onclick="oSt('"+id+"')"

const regex = /onclick=["']([a-zA-Z]+)\(['"]([^\)]+?)['"]\)(?:,\s*['"]([^'"\)]+?)['"])?['"]>/g;

c = c.replace(regex, (fullMatch, func, arg1, arg2) => {
    // Found a broken onclick, fix it
    fixes++;
    if (arg2) {
        // Two-arg function: cSt('"+id+"','status')
        return `onclick='${func}("'+arg1+'","'+arg2+'")'>`;
    } else {
        // One-arg function: oSt('"+id+"')  
        return `onclick='${func}("'+arg1+'")'>`;
    }
});

console.log(`Fixed ${fixes} onclick patterns`);

// Also fix any remaining onclick=\\\" or raw onclick=" inside h+=" strings
for (let pass = 0; pass < 3; pass++) {
    // Pattern: onclick=\"  (backslash-quote, the original problem)
    if (c.includes('onclick=\\"')) {
        // Find the context and fix
        let idx;
        while ((idx = c.indexOf('onclick=\\"')) !== -1) {
            // Look for the pattern: onclick=\\\"func('"+...+"')\\"> or similar
            let end = c.indexOf('">', idx);
            if (end === -1) break;
            let seg = c.substring(idx, end + 2);
            console.log(`  Pass ${pass+1}: fixing ${seg.substring(0, 60)}`);
            // Replace the backslash-quote with regular quote (since we'll fix quoting properly)
            c = c.substring(0, idx) + 'onclick="' + c.substring(idx + 9);
            fixes++;
        }
    }
}

// Now: some patterns might have mixed quoting after fixes
// Let me check for onclick=" inside strings that ALSO start with h+='
// Those should use escaped quotes

// Check for h+='<div ... onclick="  - this is OK
// Check for h+="<div ... onclick='  - this is also OK

// The key verification: extract JS and try to parse it
fs.writeFileSync(p, c, 'utf8');

const c2 = fs.readFileSync(p, 'utf8');
const m = c2.match(/<script>([\s\S]*?)<\/script>/);
if (m) {
    try {
        new Function(m[1]);
        console.log('✅ JS SYNTAX VALID!');
    } catch(e) {
        console.log('❌ JS ERROR: ' + e.message);
        const match = e.message.match(/line (\d+)/);
        if (match) {
            const idx = parseInt(match[1]) - 1;
            const jsLines = m[1].split('\n');
            if (jsLines[idx]) {
                console.log('  Line ' + match[1] + ': ' + jsLines[idx].substring(0, 200));
            }
        }
    }
}

console.log(`\nFile size: ${fs.statSync(p).size} bytes`);

File: _fix_v323_final.py
# -*- coding: utf-8 -*-
"""V323 Final Fix — precise byte-level replacement"""
import os, re

path = r'C:\Users\Server\.openclaw\workspace\Shop_vnext\src\Index.html'

with open(path, 'rb') as f:
    raw = f.read()

# From hex dump of line 178:
# onclick = 6f 6e 63 6c 69 63 6b
# = 3d
# \" = 5c 22  <- THIS is the problem!
# oSt(' = 6f 53 74 28 27
# "+id+" = 22 2b 69 64 2b 22
# ')\">" = 27 29 5c 22 ...
#
# The 5c 22 sequence means backslash + quote.
# In JS: h+="<div onclick=\"oSt('"+id+"')\">"
# The JS parser: string starts at ", then sees \", produces literal "
# BUT then +id+ continues... wait, let me parse carefully:
#
# The FULL line is (from hex):
# 20 20 68 2b 3d 22 ... 3d 5c 22 6f 53 74 ...
#   = "  h  +  =  "  ...  =  \  "  o  S  t ...
#
# So h += " starts a string. The string reads until the next unescaped ".
# The \ " is escaped, so it CONTINUES the string with a literal "
# Then: oSt('"+id+"')\"> 
# Wait... the second " in "+id+" would END the string!
# Because after the first \" (literal "), the next 22 (" from "+id+") ends it!
# So: string = '<div class=\'b2 bb\' onclick="'
# Outside string: "+id+"')\">เปลี่ยนสถานะ</div>";
# The " is concat, id is identifier, then + ends concat, then "+id+"')\">  
# Hmm, it's actually: '+id+"' -- the first " ends the string,
# then +id+ is expression, then "' is...
#
# Let me just fix all the broken lines by replacing the raw bytes.

BACKSLASH = b'\\'  # raw backslash byte
QUOTE = b'"'       # raw double-quote byte
SQUOTE = b"'"      # raw single-quote byte

# Pattern: onclick=\"func('"+VAR+"')\">"
# Bytes: ...onclick=" oSt(' "+id+" ')" ...
# where \" = backslash + quote = BS+DQ

# Build the broken patterns as raw bytes
def broken(func, varExpr):
    """Build the broken onclick pattern as bytes"""
    # onclick=\"func('"+VAR+"')\">
    return b'onclick=' + BACKSLASH + QUOTE + func.encode() + b'(' + SQUOTE + QUOTE + b'+' + varExpr + b'+' + QUOTE + SQUOTE + b')' + BACKSLASH + QUOTE + b'">'

def fixed(func, varExpr):
    """Build the fixed onclick pattern as bytes"""
    # onclick='func("'+VAR+'")'>
    return b"onclick='" + func.encode() + b'(' + QUOTE + SQUOTE + b'+' + varExpr + b'+' + SQUOTE + QUOTE + b')' + SQUOTE + b"'>"

fixes = [
    (broken('oSt', b'id'), fixed('oSt', b'id'), 'oSt'),
    (broken('eJob', b'id'), fixed('eJob', b'id'), 'eJob'),
    (broken('addNote', b'id'), fixed('addNote', b'id'), 'addNote'),
    (broken('window.open', b'j.folder'), fixed('window.open', b'j.folder'), 'window.open'),
]

count = 0
for old, new, label in fixes:
    if old in raw:
        raw = raw.replace(old, new)
        count += 1
        print(f"  FIXED: {label}")
    else:
        print(f"  NOT FOUND: {label}")
        # Debug: show what's near the function name
        kw = func_name = label.replace('window.open', 'window').encode()
        if kw in raw:
            idx = raw.find(kw)
            before = raw[max(0,idx-15):idx]
            after = raw[idx:idx+20]
            print(f"    Found {kw} at {idx}")
            print(f"    Before: {before}")
            print(f"    After: {after}")

# Fix addQuickNote2 - different pattern (no backslash, just raw ")
# onclick="addQuickNote2()"> inside h+="..."
old_qn = b'onclick="addQuickNote2()">'
new_qn = b"onclick='addQuickNote2()'>"
if old_qn in raw:
    raw = raw.replace(old_qn, new_qn)
    count += 1
    print(f"  FIXED: addQuickNote2")
else:
    # Check if already fixed
    if b"onclick='addQuickNote2()'>" in raw:
        print(f"  OK: addQuickNote2 already fixed")
    else:
        print(f"  NOT FOUND: addQuickNote2")

# Save
with open(path, 'wb') as f:
    f.write(raw)

print(f"\nTotal fixes: {count}")
print(f"File size: {os.path.getsize(path)} bytes")

# Verify
with open(path, 'rb') as f:
    text = f.read().decode('utf-8')

lines = text.split('\n')
bad = 0
for i, line in enumerate(lines, 1):
    if 'onclick=\\' + '"' in line and 'h+="<' in line and i >= 170 and i <= 190:
        # Check: does this have h+=" followed by ... onclick=\" (which in the string is onclick=literal"
        if 'h+="' in line:
            print(f"  ❌ POSSIBLY BROKEN line {i}: {line[:100]}")
            bad += 1

print(f"\n  Lines checked for issues. Potentially broken: {bad}")
if bad == 0:
    print("  🎉 ALL CHECKS PASSED!")
